<!doctype html> <html lang=en > <meta charset=UTF-8 > <meta name=viewport  content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv=x-ua-compatible  content="ie=edge"> <meta http-equiv=content-type  content="text/html; charset=utf-8" /> <meta name=author  content="Jeff Bezanson, Stefan Karpinski, Viral Shah, Alan Edelman, et al." /> <meta name=description  content="The official website for the Julia Language. Julia is a language that is fast, dynamic, easy to use, and open source. Click here to learn more."/> <meta property="og:title" content="The Julia Language"/> <meta property="og:image" content="/assets/images/julia-open-graph.png"/> <meta property="og:description" content="Official website for the Julia programming language"/> <link href="https://fonts.googleapis.com/css?family=Roboto:400,400i,500,500i,700,700i" rel=stylesheet > <link rel=stylesheet  href="/libs/bootstrap/bootstrap.min.css" /> <link rel=stylesheet  href="/css/app.css" /> <link rel=stylesheet  href="/css/fonts.css" /> <link rel=stylesheet  href="/css/franklin.css" /> <script async defer src="/libs/buttons.js"></script> <!-- --> <script type="application/javascript"> var doNotTrack = false; if (!doNotTrack) { window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date; ga('create', 'UA-28835595-1', 'auto'); ga('send', 'pageview'); } </script> <script async src='https://www.google-analytics.com/analytics.js'></script> <link rel=icon  href="/assets/infra/julia.ico"> <title>BioJulia Project in 2016</title> <style> .container ul li p {margin-bottom: 0;} </style> <div class="container py-3 py-lg-0"> <nav class="navbar navbar-expand-lg navbar-light bg-light" id=main-menu > <a class=navbar-brand  href="/" id=logo > <img src="/assets/infra/logo.svg" height=55  width=85  alt="JuliaLang Logo"/> </a> <button class="navbar-toggler ml-auto hidden-sm-up float-xs-left" type=button  data-toggle=collapse  data-target="#navbarSupportedContent" aria-controls=navbarSupportedContent  aria-expanded=false  aria-label="Toggle navigation"> <span class=navbar-toggler-icon ></span> </button> <div class="collapse navbar-collapse" id=navbarSupportedContent > <ul class="navbar-nav mr-auto"> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="/downloads/">Download</a> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="https://docs.julialang.org">Documentation</a> <li class="nav-item active flex-md-fill text-md-center"> <a class=nav-link  href="/blog/">Blog</a> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="/community/">Community</a> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="/learning/">Learning</a> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="/research/">Research</a> <li class="nav-item flex-md-fill text-md-center"> <a class=nav-link  href="/jsoc/">JSoC</a> <li class="nav-item donate flex-md-fill text-md-center"> <a class=github-button  href="https://github.com/sponsors/julialang" data-icon=octicon-heart  data-size=large  aria-label="Sponsor @julialang on GitHub">Sponsor</a> </ul> </div> </nav> </div> <br><br> <div class="container blog-title"> <h1>BioJulia Project in 2016</h1> <h3> <span style="font-weight: lighter;"> 30 April 2016 </span> | <span style="font-weight: bold;"></span> <span style="font-weight: bold;">Kenta Sato </span> </h3> </div> <div class=container > <p>I am pleased to announce that the next phase of BioJulia is starting&#33; In the next several months, I&#39;m going to implement many crucial features for bioinformatics that will motivate you to use Julia and BioJulia libraries in your work. But before going to the details of the project, let me briefly introduce you what the BioJulia project is. This project is supported by the Moore Foundation and the Julia project.</p> <p><a href="https://github.com/BioJulia">The BioJulia project</a> is a collaborative open source project to create an infrastructure for bioinformatics in the Julia programming language. It aims to provide fast and accessible software libraries. Julia&#39;s Just-In-Time &#40;JIT&#41; compiler enables this greedy goal without resorting to other compiled languages like C/C&#43;&#43;. The central package developed under the project is <a href="https://github.com/BioJulia/Bio.jl">Bio.jl</a>, which provides fundamental features including biological symbols/sequences, file format parsers, alignment algorithms, wrappers for external softwares, etc. It also supports several common file formats such as FASTA, FASTQ, BED, PDB, and so on. Last year I made the <a href="https://github.com/BioJulia/FMIndexes.jl">FMIndexes.jl</a> package to build a full-text search index for large genomes as a Julia Summer of Code &#40;JSoC&#41; student, and we released the first development version of Bio.jl. While the BioJulia project is getting more active and the number of contributors are growing, we still lack some important features for realistic applications. Filling in gaps between our current libraries and actual use cases is the purpose of my new project.</p> <p>So, what will be added in it? Here is the summary of my plan:</p> <ul> <li><p>Sequence analysis:</p> <ul> <li><p>Online sequence search algorithms</p> <li><p>Data structure for reference genomes</p> <li><p>Error-correcting algorithms for DNA barcodes</p> <li><p>Parsers for BAM and CRAM file formats</p> </ul> <li><p>Integration with data viewers and databases:</p> <ul> <li><p>Genome browser backend</p> <li><p>Parsers for GFF3 and VCF/BCF</p> <li><p>Database access through web APIs</p> </ul> </ul> <p>These things are of crucial importance for writing analysis programs because they connects software components &#40;e.g. programs, archives, databases, viewers, etc.&#41;; data analysis softwares in bioinformatics usually read/write formatted data from/to each other. The figure below shows common workflow of detecting genetic variants; underlined deliverables will connect softwares, archives and databases so that you can write your analysis software in the Julia language.</p> <p><img src="/assets/blog/2016-04-30-biojulia/schema.png" alt=schema  /></p> <h2 id=sequence_analysis ><a href="#sequence_analysis">Sequence Analysis</a></h2> <p>The online sequence search algorithms will come with three flavors: exact, approximate, and regular expression search algorithms. The exact sequence search literally means finding exactly matching positions of a query sequence in another sequence. The approximate search is similar to the exact search but allows up to a specified number of errors: mismatches, insertions, and deletions. The regular expression search accepts a query in regular expression, which enables flexible description of a query pattern like <a href="https://en.wikipedia.org/wiki/Sequence_motif">motifs</a>. For these algorithms, there are already half-done pull requests I&#39;m working on: <a href="https://github.com/BioJulia/Bio.jl/pull/152">#152</a>, <a href="https://github.com/BioJulia/Bio.jl/pull/153">#153</a>, <a href="https://github.com/BioJulia/Bio.jl/pull/143">#143</a>.</p> <p>After the last release of Bio.jl v0.1.0, the sequence data structure has been significantly rewritten to make biological sequence types coherent and extensible. But because we chose an encoding that requires 4 bits per base to represent DNA sequences, the DNA sequence type consumes too much memory than necessary to store a <a href="https://en.wikipedia.org/wiki/Reference_genome">reference genome</a>, which is usually composed of four kinds of DNA nucleotides &#40;denoted by A/C/G/T&#41; and &#40;consecutive and relatively small number of&#41; undetermined nucleotides &#40;denoted by N&#41;. After trying some data structures, I found that memory space of N positions can be dramatically saved using <a href="https://github.com/BioJulia/IndexableBitVectors.jl">IndexableBitVectors.jl</a>, which is a package I created in <a href="/blog/2015/10/biojulia-sequence-analysis/">JSoC 2015</a>. I&#39;m developing a separated package for reference genomes, <a href="https://github.com/BioJulia/ReferenceSequences.jl">ReferenceSequences.jl</a>, and going to improve the functionality and performance to handle huge genomes like the human genome.</p> <p>If you are a researcher or an engineer who handles high-throughput sequencing data, BAM and CRAM parsers would be the most longing feature addition in the list. BAM is the <em>de facto</em> standard file format to accommodate aligned sequences and most sequence mappers generate alignments in this format. CRAM is a storage-efficient alternative of BAM and is getting popular reflecting explosion of accumulated sequence data. Since these files contain massive amounts of DNA sequences from high-throughput sequencing machines, high-speed parsing is a practically desirable feature. I&#39;m going to concentrate on the speed by careful tuning and multi-thread parallel computation which is planned to be introduced in the next Julia release.</p> <h2 id=integration_with_data_viewers_and_databases ><a href="#integration_with_data_viewers_and_databases">Integration with Data Viewers and Databases</a></h2> <p>Genome browsers enable to interactively visualize genetic features found in individuals and/or populations. For example, using the UCSC Genome Browser, you can investigate genetic regions along with sequence annotations <a href="https://genome.ucsc.edu/cgi-bin/hgTracks?db&#61;hg38&amp;lastVirtModeType&#61;default&amp;lastVirtModeExtraState&#61;&amp;virtModeType&#61;default&amp;virtMode&#61;0&amp;nonVirtPosition&#61;&amp;position&#61;chr9&#37;3A133206569-133324246&amp;hgsid&#61;491214673_Ob3A4L4zTLibsCuyq7xgazU3Goqg">around the ABO gene</a> in a window. Genome browser is one of the most common visualizations and hence lots of softwares have been developed but unfortunately there is no standardized interface. So, we will need to select a promising one that is an open source and supporting interactions with other softwares. The first candidate is <a href="https://jbrowse.org/">JBrowse</a>, which is built with modern JavaScript and HTML5 technologies. It also supports RESTful APIs and hence it can fetch data from a backend server via HTTP. I&#39;m planning to make an API server that responds to queries from a genome browser to interactively visualize in-memory data.</p> <p>Many databases distribute their data in some standardized file formats. As for genetic annotations and variants, GFF3 and VCF would be the most common formats. If you are using data from human or mouse, you should know various annotations are available from <a href="https://www.gencodegenes.org/">the GENCODE project</a>. It offers data in GTF or GFF3 file formats. NCBI provides human variation sets in VCF file formats <a href="https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/">here</a>. These file formats are text, so you may think it is trivial to write parsers when you need them. It is partially true — if you don&#39;t care about completeness and performance. Parsing a text file format in a naive way &#40;for example, <code>split</code> a line by a tab character&#41; allocates many temporary objects and often leads to degrade performance, while careful tuning of a parser leads to complicated code that is hard to maintain. <a href="https://github.com/dcjones">@dcjones</a> challenged this problem and made a great work and made Julia support for <a href="https://www.colm.net/open-source/ragel/">Ragel</a>, which generates Julia code that executes a finite state machine. Daniel&#39;s talk of the JuliaCon 2015 is helpful to know about the details if you are interested:</p> <iframe width=560  height=315  src="https://www.youtube.com/embed/sQvNNj3MthQ" frameborder=0  allowfullscreen></iframe> <p>Sometimes you may need only a part of data provided by a database. In such a case, web-based APIs are handy to fetch necessary data on demand. <a href="http://www.biomart.org/">BioMart Central Portal</a> offers a unified access point to a range of biological databases that is programmatically accessible via REST and SOAP APIs. Julian wrapper to BioMart will make it much easier to access data by automatically converting response to Julia objects. In the R language, the <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package is one of the most downloaded packages in Bioconductor packages: <a href="https://www.bioconductor.org/packages/stats/">https://www.bioconductor.org/packages/stats/</a>.</p> <h2 id=try_biojulia ><a href="#try_biojulia">Try BioJulia&#33;</a></h2> <p>We need users and collaborators of our libraries. Feedbacks from users in the real world are the most precious thing to improve the quality of our libraries. We welcome feature requests and discussions that will make bioinformatics easier and faster. Tools for phylogenetics and structural biology, which I didn&#39;t mention in this post, are also under active development. You can post issues here: <a href="https://github.com/BioJulia/Bio.jl/issues">https://github.com/BioJulia/Bio.jl/issues</a>; if you want to get in touch with us more casually, this Gitter room may be more convenient: <a href="https://gitter.im/BioJulia/Bio.jl">https://gitter.im/BioJulia/Bio.jl</a>.</p> </div> <br><br> <footer class="container-fluid footer-copy"> <div class=container > <div class=row > <div class="col-md-10 py-2"> <p>This website is built with <a style="color: #7a95dd" href="https://franklinjl.org">Franklin.jl</a> - a Julia native package for building websites. We thank <a style="color: #7a95dd" href="https://www.fastly.com">Fastly</a> for their generous infrastructure support. <p>©2020 JuliaLang.org <a style="color: #7a95dd" href="https://github.com/JuliaLang/www.julialang.org/graphs/contributors">contributors</a>. The content on this website is made available under the <a style="color: #7a95dd" href="https://github.com/JuliaLang/www.julialang.org/blob/master/LICENSE.md">MIT license</a>. </div> <div class="col-md-2 py-2"> <a class=github-button  href="https://github.com/sponsors/julialang" data-icon=octicon-heart  data-size=large  aria-label="Sponsor @julialang on GitHub">Sponsor</a> </div> </div> </div> </footer> <script src="/libs/jquery/jquery.min.js"></script> <script src="/libs/bootstrap/bootstrap.min.js"></script> <script src="/libs/platform.js"></script>