---
type: "article"
title: "Simultaneous identification of specifically interacting paralogs and interprotein contacts by direct coupling analysis"
authors:
- Gueudré, Thomas
- Baldassi, Carlo
- Zamparo, Marco
- Weigt, Martin
- Pagnani, Andrea
year: "2016"
journal: "Proceedings of the National Academy of Sciences"
volume: "113"
number: "43"
pages: "12186--12191"
doi: "10.1073/pnas.1607570113"
packages:
  ParalogMatching.jl: https://github.com/Mirmu/ParalogMatching.jl
---
Understanding protein−protein interactions is central to our understanding of almost all complex biological processes. Computational tools exploiting rapidly growing genomic databases to characterize protein−protein interactions are urgently needed. Such methods should connect multiple scales from evolutionary conserved interactions between families of homologous proteins, over the identification of specifically interacting proteins in the case of multiple paralogs inside a species, down to the prediction of residues being in physical contact across interaction interfaces. Statistical inference methods detecting residue−residue coevolution have recently triggered considerable progress in using sequence data for quaternary protein structure prediction; they require, however, large joint alignments of homologous protein pairs known to interact. The generation of such alignments is a complex computational task on its own; application of coevolutionary modeling has, in turn, been restricted to proteins without paralogs, or to bacterial systems with the corresponding coding genes being colocalized in operons. Here we show that the direct coupling analysis of residue coevolution can be extended to connect the different scales, and simultaneously to match interacting paralogs, to identify interprotein residue−residue contacts and to discriminate interacting from noninteracting families in a multiprotein system. Our results extend the potential applications of coevolutionary analysis far beyond cases treatable so far.